File: go/algorithms/tim.go
package algorithms

const minRun = 32

// TimSort sorts in-place using a simplified TimSort.
func TimSort(arr []int) {
    n := len(arr)
    for i := 0; i < n; i += minRun {
        right := i + minRun
        if right > n {
            right = n
        }
        insertionRange(arr, i, right)
    }
    for size := minRun; size < n; size *= 2 {
        for left := 0; left < n; left += 2 * size {
            mid := left + size
            right := left + 2*size
            if mid > n {
                mid = n
            }
            if right > n {
                right = n
            }
            if mid < right {
                mergeInPlace(arr, left, mid, right)
            }
        }
    }
}

func insertionRange(arr []int, l, r int) {
    for i := l + 1; i < r; i++ {
        key := arr[i]
        j := i - 1
        for j >= l && arr[j] > key {
            arr[j+1] = arr[j]
            j--
        }
        arr[j+1] = key
    }
}

func mergeInPlace(arr []int, l, m, r int) {
    left := append([]int(nil), arr[l:m]...)
    right := append([]int(nil), arr[m:r]...)
    i, j, k := 0, 0, l
    for i < len(left) && j < len(right) {
        if left[i] <= right[j] {
            arr[k] = left[i]
            i++
        } else {
            arr[k] = right[j]
            j++
        }
        k++
    }
    for i < len(left) {
        arr[k] = left[i]
        i++
        k++
    }
    for j < len(right) {
        arr[k] = right[j]
        j++
        k++
    }
}

File: rust/src/tim.rs
pub fn tim_sort(arr: &mut [i32]) {
    const MIN_RUN: usize = 32;
    let n = arr.len();
    let mut i = 0;
    while i < n {
        let right = usize::min(i + MIN_RUN, n);
        insertion_range(arr, i, right);
        i += MIN_RUN;
    }
    let mut size = MIN_RUN;
    while size < n {
        let mut left = 0;
        while left < n {
            let mid = usize::min(left + size, n);
            let right = usize::min(left + 2 * size, n);
            if mid < right {
                merge_in_place(arr, left, mid, right);
            }
            left += 2 * size;
        }
        size *= 2;
    }
}

fn insertion_range(arr: &mut [i32], l: usize, r: usize) {
    for i in l + 1..r {
        let key = arr[i];
        let mut j = i;
        while j > l && arr[j - 1] > key {
            arr[j] = arr[j - 1];
            j -= 1;
        }
        arr[j] = key;
    }
}

fn merge_in_place(arr: &mut [i32], l: usize, m: usize, r: usize) {
    let left = arr[l..m].to_vec();
    let right = arr[m..r].to_vec();
    let mut i = 0;
    let mut j = 0;
    let mut k = l;
    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            arr[k] = left[i];
            i += 1;
        } else {
            arr[k] = right[j];
            j += 1;
        }
        k += 1;
    }
    while i < left.len() {
        arr[k] = left[i];
        i += 1;
        k += 1;
    }
    while j < right.len() {
        arr[k] = right[j];
        j += 1;
        k += 1;
    }
}
